Add Google Tag Manager callback to support nightwatch e2e testing

barelyfitz opened the pull request:
Adds a callback option for the Google Tag Manager provider, to be used for verifying GTM calls in nightwatch (needed for [rent.js PR#1524](https://github.com/rentpath/rent-js/pull/1524))

lfender6445 commented on src/providers/GoogleTagManager.js:
```diff
@@ -22,7 +22,27 @@ export default class GoogleTagManager {
-    window.dataLayer.push(data)
+    const { eventCallback, eventTimeout } = this.config
+
+    const newData = { ...data }
```
is a shallow clone ok for the data were working with?

barelyfitz replied:
Probably in this case - I'm just adding a couple new properties, not modifying any of the values. I could just as easily have added those properties directly to `data` but I think eslint doesn't like that.

lfender6445 commented on src/providers/GoogleTagManager.js:
```diff
@@ -22,7 +22,27 @@ export default class GoogleTagManager {
   }
 
   track(data) {
```
is there a reason why adding the callback as an argument would not work?

barelyfitz replied:
A few reasons that might not work:
- The `track()` function is called by event-tracker, not by the application (rent-js), so at some point you would still need to configure a callback to be used
- Some providers do not support callbacks, so a Tealium call would not honor the callback like GoogleTagManager
- The callback function should be called by GoogleTagManager after it successfully makes the tagging call. That lets us verify that the tagging call was actually made to GTM (vs verifying that event-manager's `track()` function was called). Sort of a subtle difference, but important when it comes to testing.
- Because of this I thought it was best to make the callback function an option for Google Tag Manger provider, rather than an overall callback for event-tracker.

lfender6445 commented on src/providers/GoogleTagManager.js:
```diff
@@ -22,7 +22,27 @@ export default class GoogleTagManager {
+      newData.eventCallback = () => {
+        // Pass the data back to the callback so it will know
+        // what tagging was requested
+        eventCallback(data)
```
im confused

im confused - wouldnt this be

```jsx
newData.eventCallback = eventCallback(data)
```

because in the other pr you have 

```jsx
    eventCallback: data => {
      // Check for a global function gtmEventCallback which will be used
      // for nightwatch tests to verify GTM tracking calls
      if (window.gtmEventCallback) {
        window.gtmEventCallback(data)
      }
    },
```

so data.eventCallback is set to a fn 

i might be missing something

barelyfitz replied:
Oh, and it would have been simply:
```
newData.eventCallback = eventCallback
```
Except GTM doesn't pass any arguments back to the callback, and we really need to know the tagging information, so that's why we pass `data` back to our callback.

barelyfitz commented:
@lfender6445 I renamed the config properties to trackCallback and trackTimeout, hopefully that will give a little more differentiation from the GTM eventCallback properties. Also added some comments.
